feat: add schema serialization modules

rodesai opened the pull request:
### Description 

This patch adds serialization/deserialization for arbitrary logical
schemas using KSQL's grammar/parser.

The grammar has been extended to support parsing full schema specs that
include meta fields and source qualifiers.

The table element and logical schema building code has been extended to
support meta fields and source qualifiers.

Note: this PR depends on #3720, and has that as the first commit. So please
review those changes in that PR.

rodesai commented:
> When reviewing this PR it would be really useful to understand why you're making these changes. What's the end goal?
> Personally, I find the addition of methods to the logical schema builder for meta fields very worrying. The meta fields, (currently only ROWTIME), are by definition always present. So it just doesn't make conceptual sense to me that you'd need to add them in the builder.

The goal here is to support serializing/deserializing intermediate schemas in execution plans. This requires two things:
    - Allowing the schema to include source qualifiers in column names.
    - Serializing/deserializing meta columns. We could get around this by always deserializing these with meta columns, however in the case where the schema has source qualifiers coming from multiple sources (e.g. a join), we won't know the right source qualifier to use. In the end it seemed simplest/most foolproof to just allow serializing/deserializing any schema without making assumptions.

I think there's also a case to be made for not serializing the intermediate schemas, and just computing these as we go along. Then we only need to worry about serializing/deserializing meta-store schemas, which aren't source qualified. I'm leaning more in this direction now, so I'll most likely close this PR out and open one that does that (in which case we can reuse the existing serialization modules)

big-andy-coates commented:
It's an interesting one for sure!  I think serializing the schema, in some form, makes sense.

I'm guessing you're adding these so that the meta fields can be serialized with the schema into the command topic, as this will allow us to know what meta fields were available at the time the query was started.  But is this necessary?  

Sure, we don't currently expose partition, offset or headers, but if we choose to do so in the future then I think the only way an existing query would break is if the query had a column in the data with the same name.  Otherwise, if the query was run with the new metafields it would just ignore them.  

What would we do if we encountered a clash between a column name in an existing query and a new meta column?  Even knowing there was a clash, it still would not be easy to continue to process the query as before.

If this is the reason you're adding the meta columns, then I don't think we should.  Instead, we could:
1. Move `ROWTIME`, and any new meta columns we add, to be within a single `STRUCT` column called `Metadata` or `RowMetadata` etc.  (We can do this as we move to ksqlDB without it being a breaking change).  We make `Metadata` a reserved column name. We're then free to add new fields to the struct as we like without breaking backwards compatibility.
1. Alternatively, we could just reserve the names of the meta columns we're thinking of adding, i.e. `PARTITION`, `OFFSET` and `HEADERS`. However, we'll still potentially run into issues if we add something not already reserved.
1. Failing that, we could store the meta columns in the payload we save to the command topic, just not inside the schema, as we want the schema to strongly enforce the implicit meta columns.  Though, as mentioned above, I'm not even sure this would help should we run into a name clash.

What do you think?

Here's some food for thought...

### SELECT * and meta columns
In a join, when we have the meta columns coming from both sides, e.g. `SELECT * FROM A JOIN B ...`, we end up with `A_ROWTIME` and `B_ROWTIME` style columns in the result schema. However, as these still meta columns?  Personally, I think not. These are now just normal `VALUE` columns, where the value comes from the `META` columns in the source.

If you extrapolate the reasoning in the above paragraph to non-joins, then it would seem to follow that doing a `SELECT * FROM A` that the result schema should also have a `A_ROWTIME` `VALUE` column. However, this doesn't feel nice.

With ksqlDB being released we're at a perfect time to change the semantics of such queries.

Maybe we could say that meta columns are not included in a `select *`.  If you want them in your output then you should explicitly add them.  

(At the same time we should exclude `ROWKEY` from the join schema, given we join on it anyway I don't see the value in adding `A_ROWKEY` and `B_ROWKEY`, given the values will match and will match the actual value in the key: https://github.com/confluentinc/ksql/issues/3731)

#### Schema of Kafka Streams topology nodes

At the moment we use `LogicalSchema` to represent the schema at each node of the KS topology. As stated above, this currently includes implicit metadata columns. However, I'm not sure this makes sense.  Specifically, think about the schema post an aggregation:

```sql
SELECT ID, COUNT() FROM X GROUP BY ID;
```

Say we've introduced a `PARTITION` meta column, or `OFFSET`. Does it actually make sense for the aggregate result schema to include these?  Personally, I think not.  The source's `PARTITION` and `OFFSET` meta columns should be lost during the aggregation.  Arguably, the same should be true of `ROWTIME`.  It shouldn't be copied across the aggregation, but currently is.

Interestingly, this leans me in favour of your change to add meta columns to the schema builder.  Maybe the schema used for data sources should explicitly have the current meta columns added. While the result schema of the aggregate would _not_ have any implicit meta columns at all.

rodesai commented:
> Say we've introduced a PARTITION meta column, or OFFSET. Does it actually make sense for the aggregate result schema to include these? Personally, I think not. The source's PARTITION and OFFSET meta columns should be lost during the aggregation. Arguably, the same should be true of ROWTIME. It shouldn't be copied across the aggregation, but currently is.

I agree for partition and offset. I think accessing the row time makes sense for the result of an aggregate, but should be the timestamp fo the result (max of the record and stored timestamp), not what it is today. This could be useful if we were to support nested queries in FROM, e.g. `SELECT ROWTIME, ..., FROM (SELECT ... FROM ...);

rodesai commented:
> Maybe we could say that meta columns are not included in a select *. If you want them in your output then you should explicitly add them.

I agree here. We actually already do this for `select *` from a stream/table.

rodesai commented:
I think the discussion above raises a few orthogonal points/questions:

- How do we represent and support querying metadata? Should these even be in the stream/table schema?
- What metadata should be available at different points in the query?
- What should the semantics be for '*' with respect to metadata columns?

Definitely important to discuss, but probably better done under https://github.com/confluentinc/ksql/issues/3734

For this PR, I don't think I actually need the meta fields to be serialized or to be supported by the builder. The main reason I put them in was because I didn't know which source qualifier to use for it for joins. Ultimately I think it makes more sense to compute the schema as we build the streams app from the plan, and only write it in the plan for the source and at points where we write to Kafka. This gives us a bit more flexibility to change stuff without having to iterate step versions. To get this patch unblocked I'll just pick the first seen qualifier for ROWTIME when building schemas from the builder.

big-andy-coates commented:
> For this PR, I don't think I actually need the meta fields to be serialized or to be supported by the builder. The main reason I put them in was because I didn't know which source qualifier to use for it for joins. But I don't think the meta field is even used anywhere there is a source qualifier, so it shouldn't matter. Ultimately I think it makes more sense to compute the schema as we build the streams app from the plan, and only write it in the plan for the source and at points where we write to Kafka. This gives us a bit more flexibility to change stuff without having to iterate step versions. In either case there's no qualifiers though, so more reason to not worry about the qualifier on rowtime.

Cool. Can this PR be closed then?

rodesai commented:
> Cool. Can this PR be closed then?

Not quite. I've moved the schema deserializer code so that it's accessible from the parser/engine/execution, and added support for specifying format options when serializing. I've also added support for serializing/deserializing qualified schemas, which I don't think I'll need once klip-6 is done, but will help me get some stuff unblocked in the meanwhile.

big-andy-coates commented on ksql-common/src/main/java/io/confluent/ksql/json/KsqlTypesSerializationModule.java:
```diff
@@ -16,11 +16,15 @@
-
   public KsqlTypesSerializationModule() {
-    addSerializer(LogicalSchema.class, new LogicalSchemaSerializer());
+    this(FormatOptions.none());
```
I'd remove the default constructor so people have to make a conscious choice when to pass as the format options.

big-andy-coates commented:
OK, give me a ping once  #3720 is merged and I'll take another look

big-andy-coates commented on ksql-parser/src/main/java/io/confluent/ksql/parser/tree/TableElement.java:
```diff
@@ -63,11 +65,29 @@ public TableElement(
+      final Namespace namespace,
+      final ColumnName name,
+      final Type type,
+      final Optional<SourceName> source
```
`TableElement` is the pojo that elements in the CT/CS schema are deserialized to, right?  e.g. `CREATE TABLE FOO (X INT, Y DOUBLE) ...` would result in a two `TableElement`s for `X` and `Y`.

It feels wrong to extend the model pojo to hold things that our syntax doesn't need.

Can you explain _why_ you need an optional source name, because at the moment I don't understand the reasoning.

big-andy-coates replied:
From an offline chat, this is needed because the `LogicalSchemaDeserializer` uses the `SchemaParser` which returns `TableElements`.  

I think a better approach would be to keep `TableElement` as it is, (with no source), as that is what the contract is for CS/CT statements. Instead, add a new type to capture Schema elements.

big-andy-coates commented on ksql-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4:
```diff
@@ -86,12 +86,24 @@ resultMaterialization
 
 tableElement
-    : identifier type (KEY)?
+    : identifier type
```
Do we need to drop `(KEY)?` from `tableElement`? Can we not just have:

```
schemaElements
    : '(' schemaElement (',' schemaElement)* ')'
    ;
    
schemaElement
    : (identifier '.')? tableElement
    ;

tableElements
   : '(' tableElement (',' tableElement)* ')'
   
tableElement
   : identifier type (KEY)?
```

And have both `visitStatementTableElement` and `visitTableElement` in the `AstBuilder`?

big-andy-coates commented on ksql-parser/src/main/java/io/confluent/ksql/parser/SchemaParser.java:
```diff
@@ -83,16 +86,31 @@ public void syntaxError(
     final SqlTypeParser typeParser = SqlTypeParser.create(typeRegistry);
 
-    final List<TableElement> elements = parser.tableElements().tableElement()
+    final List<TableElement> elements = parser.schemaTableElements().schemaTableElement()
```
It might be worth splitting `SchemaParser` into more strongly typed `LogicalSchemaParser` and `TableElementsParser`

big-andy-coates commented on ksql-parser/src/main/java/io/confluent/ksql/parser/tree/TableElement.java:
```diff
@@ -38,6 +39,7 @@
   private final Namespace namespace;
   private final ColumnName name;
   private final Type type;
+  private final Optional<SourceName> source;
```
If we're adding a new field it should be part of `equals`, `hashCode` and `toString`.

big-andy-coates commented on ksql-parser/src/main/java/io/confluent/ksql/parser/tree/TableElements.java:
```diff
@@ -111,12 +126,14 @@ private TableElements(final ImmutableList<TableElement> elements) {
         .filter(e -> e.getValue() > 1)
         .map(Entry::getKey)
-        .map(ColumnName::toString)
+        .map(l -> String.format("%s.%s %s", l.get(0), l.get(1), l.get(2)))
```
This wouldn't be needed if `ColumnName` took the new source name into account.

big-andy-coates commented on ksql-parser/src/test/java/io/confluent/ksql/parser/json/LogicalSchemaDeserializerTest.java:
```diff
@@ -83,10 +86,43 @@ public void shouldDeserializeSchemaWithKeyAfterValue() throws Exception {
+    final LogicalSchema schema = MAPPER.readValue(json, LogicalSchema.class);
+
+
+    Assert.assertThat(schema, is(LogicalSchema.builder()
```
nit: use hamcreast `assertThat`.

big-andy-coates commented on ksql-parser/src/test/java/io/confluent/ksql/parser/json/LogicalSchemaDeserializerTest.java:
```diff
@@ -83,10 +86,43 @@ public void shouldDeserializeSchemaWithKeyAfterValue() throws Exception {
+    // When:
+    final LogicalSchema schema = MAPPER.readValue(json, LogicalSchema.class);
+
+
```
nit: missing `// Then:` (same below)

big-andy-coates commented on ksql-parser/src/test/java/io/confluent/ksql/parser/json/LogicalSchemaDeserializerTest.java:
```diff
@@ -83,10 +86,43 @@ public void shouldDeserializeSchemaWithKeyAfterValue() throws Exception {
+    final LogicalSchema schema = mapper.readValue(json, LogicalSchema.class);
+
+
+    Assert.assertThat(schema, is(LogicalSchema.builder()
```
I think this test would be more robust if it actually explicitly checked for the presence of the meta columns, e.g. 

```java
assertThat(schema.meta(), is(not(empty()));
```

It's also more readable as to what the test is testing.
```

big-andy-coates commented on ksql-parser/src/test/java/io/confluent/ksql/parser/tree/TableElementsTest.java:
```diff
@@ -112,26 +115,6 @@ public void shouldThrowOnDuplicateValueColumns() {
   }
 
-  @Test
-  public void shouldThrowOnDuplicateKeyValueColumns() {
```
Why has this test been removed?  Bad things happen if we allow duplicates at the moment. We really don't want to allow this!

If this is a requirement for the parsing of logical schemas, then it once again suggests that we need two different parsers: one of logicl schemas and one for CT/CS schemas.

rodesai commented:
Closing in favor of #3845
